Enable codegen of per-dispatch key derivative formulas in derivatives.yaml

mikaylagawarecki opened the pull request:
`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  Default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82916
* __->__ #82801
* __->__ #82801

mikaylagawarecki committed:
[WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"

`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


* __->__ #82801

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[DO NOT REVIEW] [WIP] Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"

`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


* __->__ #82801

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"

`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


* __->__ #82801

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -12273,6 +12273,21 @@
+    CompositeExplicitAutograd: _test_autograd_multiple_dispatch
+    # this is necessary because NestedTensor has ben removed from the CompositeExplicitAutograd keyset
+    # see Note [NestedTensor Not Included in Backend Keys]
+    NestedTensorCPU, NestedTensorCUDA: _test_autograd_multiple_dispatch
```
nit: This can be added on the same line as the CompositExplicitAutograd above.

bhosmer replied:
...but (just to state the obvious) definitely keep the comment, i.e. "the NestedTensor keys are necessary because..."

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -12273,6 +12273,21 @@
+    # this is necessary because NestedTensor has ben removed from the CompositeExplicitAutograd keyset
+    # see Note [NestedTensor Not Included in Backend Keys]
+    NestedTensorCPU, NestedTensorCUDA: _test_autograd_multiple_dispatch
+  tags: view_copy
```
This is not a copy variant of a view right?

albanD commented on test/test_autograd.py:
```diff
@@ -8887,6 +8887,102 @@ def fn(x1, x2):
         torch.autograd.gradcheck(fn, [inp_c, inp_r], check_forward_ad=True)
 
+class TestAutogradMultipleDispatch(TestCase):
+    def test_autograd_multiple_dispatch_registrations(self, device):
```
This is a good way to check this

albanD commented on test/test_autograd.py:
```diff
@@ -8887,6 +8887,102 @@ def fn(x1, x2):
+            # bogus gradient registered for AutogradCUDA is grad * 2
+            self.assertEqual(t.grad, grad * 2)
+
+        # test registered AutogradNestedTensor formula
```
This is a duplicate of `test_autograd_composite_implicit_and_dispatch_registration` no?

mikaylagawarecki replied:
Hm not quite, I registered a different formula for `_test_autograd_multiple_dispatch.one`, the composite implicit test checks `_test_autograd_multiple_dispatch.two`. I was just trying to check that multiple registrations work in this test while the next test checks a single registration + composite implicit, I can remove this part if it's unecessary though!

albanD replied:
Ho right. I got confused by the two overloads :D

bhosmer replied:
Probably worth one-liners above the two functions noting the distinction (basically just the content in your reply to Alban)

albanD commented on test/test_autograd.py:
```diff
@@ -8887,6 +8887,102 @@ def fn(x1, x2):
+        a_view_copy = torch.view_copy(a, (2, 2))
+
+        # using _test_autograd_multiple_dispatch.one
+        out_view_copy = torch._test_autograd_multiple_dispatch(a_view_copy)
```
I'm a bit confused here? The difference between these two is just that one input is a view and not the other?

mikaylagawarecki replied:
This test was wrong, I misunderstood what a view copy op is sorry I will fix it!

albanD commented on test/test_autograd.py:
```diff
@@ -8901,6 +8997,12 @@ def fn(x1, x2):
+instantiate_device_type_tests(
+    TestAutogradMultipleDispatch,
+    globals(),
+    except_for=None
```
Is this line needed?
Only should we do only CPU and CUDA for this?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2789,6 +2797,22 @@
+  dispatch:
+    AutogradNestedTensor:
+      self: grad.mul(grad).add(grad)
+  output_differentiability: [True]
```
Is this needed when there is no default?

mikaylagawarecki replied:
Nope! Will remove

albanD commented on tools/autograd/gen_trace_type.py:
```diff
@@ -458,11 +458,20 @@ def emit_trace_body(f: NativeFunction) -> List[str]:
+    # We do not append key if it is Default so that generated functions from
+    # before per-dispatch-key derivatives were added retain the same names.
+    if key != "Default":
+        name = name + f"_{key}"
```
Why do we actually change the name here? Only one function should be generated anyways and so there won't be any conflict.

mikaylagawarecki replied:
This function is used in `gen_wrapper_registration` in gen_variable_type.cpp which generates the registrations in VariableType.cpp, we need multiple functions in variableType.cpp and I'm differentiating this by appending the keys (e.g. AddTensorBackward vs AddTensorBackward_AutogradNestedTensor)

albanD replied:
Ho ok! Sounds good!

albanD commented on torchgen/api/autograd.py:
```diff
@@ -292,26 +304,28 @@ def dispatch_strategy(fn: NativeFunctionWithDifferentiabilityInfo) -> str:
     """
 
-    info_by_schema = {info.func.func: info for info in differentiability_infos}
+    info_by_schema = differentiability_infos
```
nit: not need for the extra variable

mikaylagawarecki commented on test/test_autograd.py:
```diff
@@ -8887,6 +8887,102 @@ def fn(x1, x2):
+            else:
+                torch._test_autograd_multiple_dispatch(dual_input)
+
+    def test_view_copy(self, device):
```
This test isn't testing view_copy correctly, fixing.

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -48,34 +49,49 @@
+
+    for _, info_dispatch_dict in infos.items():
+        # maybe_view_group only needs to be calculated once per info_dispatch_dict
+        maybe_view_group = None
```
Why? Also since this is just a lookup, why not keep the original code?

mikaylagawarecki replied:
I think the derivatives of view_copy oops need to be generated per dispatch key, which is why I modified the code is my understanding correct here?

albanD replied:
Sure, but you could have kept the `maybe_view_group = view_name_to_group.get(info.func.func.name, None)` call inside the inner for loop.

albanD reviewed (commented):
This looks great!
Just wanted to confirm what gets generated in `generated/Functions.cpp` after this change?

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -373,10 +373,15 @@ def gen_autograd_functions_lib(
     for each every differentiable torch function.
     """
 
+    # get a 1D list of diffinfos, we do not need them to be per FunctionSchema/DispatchKey here
```
What is the diff in Functions.cpp before/after this PR?

bhosmer replied:
+1, would be useful to link a paste of this to the PR

mikaylagawarecki replied:
@albanD here is the diff I'll have to check where the `copy_range `is coming from but essentially there are now functions `TestAutogradMultipleDispatchBackward0`, `TestAutogradMultipleDispatchBackwardAutogradNestedTensor0`, `TestAutogradMultipleDispatchBackwardAutogradCUDA0`, `TestAutogradMultipleDispatchBackwardAutogradNestedTensor1`

I am of the opinion that this is correct, the first 3 correspond to `_test_autograd_multiple_dispatch_backward.one` and the last to `_test_autograd_multiple_dispatch_backward.two` and these are the function names are the ones called in VariableType.cpp from the respective variable type functions in lines like  

`std::shared_ptr<TestAutogradMultipleDispatchBackward0> grad_fn;`

`grad_fn = std::shared_ptr<TestAutogradMultipleDispatchBackward0>(new TestAutogradMultipleDispatchBackward0(), deleteNode);`
 
 <details>
<summary>Click here for diff!</summary>
<br>

```
5368d5367
<   const auto& grad_cloned_coefficient = grads[1];
5369a5369
>   const auto& grad_cloned_coefficient = grads[1];
10029a10030,10085
>     copy_range(grad_inputs, self_ix, grad_result);
>   }
>   return grad_inputs;
> }
> variable_list TestAutogradMultipleDispatchBackward0::apply(variable_list&& grads) {
> 
> 
>   IndexRangeGenerator gen;
>   auto self_ix = gen.range(1);
>   variable_list grad_inputs(gen.size());
>   const auto& grad = grads[0];
>   bool any_grad_defined = any_variable_defined(grads);
>   if (should_compute_output({ self_ix })) {
>     auto grad_result = any_grad_defined ? (grad.expand(self_sizes) + 1) : Tensor();
>     copy_range(grad_inputs, self_ix, grad_result);
>   }
>   return grad_inputs;
> }
> variable_list TestAutogradMultipleDispatchBackwardAutogradNestedTensor0::apply(variable_list&& grads) {
> 
> 
>   IndexRangeGenerator gen;
>   auto self_ix = gen.range(1);
>   variable_list grad_inputs(gen.size());
>   const auto& grad = grads[0];
>   bool any_grad_defined = any_variable_defined(grads);
>   if (should_compute_output({ self_ix })) {
>     auto grad_result = any_grad_defined ? (grad.mul(grad)) : Tensor();
>     copy_range(grad_inputs, self_ix, grad_result);
>   }
>   return grad_inputs;
> }
> variable_list TestAutogradMultipleDispatchBackwardAutogradCUDA0::apply(variable_list&& grads) {
> 
> 
>   IndexRangeGenerator gen;
>   auto self_ix = gen.range(1);
>   variable_list grad_inputs(gen.size());
>   const auto& grad = grads[0];
>   bool any_grad_defined = any_variable_defined(grads);
>   if (should_compute_output({ self_ix })) {
>     auto grad_result = any_grad_defined ? (grad.expand(self_sizes) * 2) : Tensor();
>     copy_range(grad_inputs, self_ix, grad_result);
>   }
>   return grad_inputs;
> }
> variable_list TestAutogradMultipleDispatchBackwardAutogradNestedTensor1::apply(variable_list&& grads) {
> 
> 
>   IndexRangeGenerator gen;
>   auto self_ix = gen.range(1);
>   variable_list grad_inputs(gen.size());
>   const auto& grad = grads[0];
>   bool any_grad_defined = any_variable_defined(grads);
>   if (should_compute_output({ self_ix })) {
>     auto grad_result = any_grad_defined ? (grad.mul(grad).add(grad)) : Tensor();
```

</details>

albanD replied:
Perfect!

i wasn't too worried anyway given that the test pass ;)

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -733,9 +738,52 @@ def gen_variable_type(
+        install_dir=out + "/templates", template_dir=template_path, dry_run=False
+    )
+    fm1.write(
+        "VariableType.cpp",
```
Sounds good.

bhosmer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -12273,6 +12273,21 @@
     CompositeExplicitAutograd: _test_warn_in_autograd
 
+# Note: this function is only for testing.
+- func: _test_autograd_multiple_dispatch.one(Tensor self) -> Tensor
```
What if the overload suffixes were more descriptive of what's going on, rather than just "one" and "two"? Like, IIUC from `derivatives.yaml`, `.one` is a "full coverage" test whereas `.two` is NT only. I think maybe using something like those (or whatever the correct characterizations are) as suffixes would help readability, be a good connective thread between the different moving parts here, etc

bhosmer commented on aten/src/ATen/native/TestOps.cpp:
```diff
@@ -74,5 +74,13 @@ Tensor _test_warn_in_autograd(const Tensor &self) {
   return self.clone();
 }
 
+Tensor _test_autograd_multiple_dispatch(const Tensor &self) {
```
maybe a quick comment here about what these are for, how the real action is in derivatives.yaml, something like that 😁  (comment above the `_test_ambiguous_defaults` overloads is basically what I'm thinking of)

bhosmer commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -417,9 +422,14 @@ def gen_autograd_functions_python(
+        for diffinfo_dict in differentiability_infos.values()
+        for info in diffinfo_dict.values()
+    ]
+    infos = list(filter(lambda info: info.args_with_derivatives, diff_info_list))
```
any way this could be factored out into a helper and called from here and 378 as well? (assuming it doesn't disappear per Alban's comment)

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -733,9 +738,52 @@ def gen_variable_type(
+        return "\n\n".join(library_impl_macro_list)
+
+    # Generate a new template from VariableType.cpp which replaces ${wrapper_registrations}
+    # with per key TORCH_LIBRARY_IMPL macros for each key that appears in derivatives.yaml
```
I'm wondering if there isn't a more straightforward way to do this, like a new macro that does the necessary expansion directly rather than generating a new template that's then re-expanded. But I might be overlooking some difficulty. (maybe an example of the generated template would help clarify, if it's easy to throw in a paste)

mikaylagawarecki replied:
Generated template is pasted in the dropdown below! 
There will be a separate 
```
TORCH_LIBRARY_IMPL(aten, {key}, m) {
${wrapper_registrations_{key}}
}
``` 
for each key that appears in derivatives.yaml! 
<details>
<summary>Here's the generated template</summary>
<br>

```
#include "torch/csrc/autograd/VariableTypeUtils.h"
#include "torch/csrc/autograd/generated/VariableType.h"
#include "torch/csrc/autograd/FunctionsManual.h"

#include <ATen/RedispatchFunctions.h>
#include <ATen/core/TorchDispatchModeTLS.h>
#include <torch/library.h>


// @generated by torchgen/gen.py from VariableType.cpp

// NOTE [Sharded File]: on this file's split-into-shards state
//
// Back in the good old days, VariableType.cpp was generated as one
// file with every function in it, and everything was great and
// simple.
//
// However, this file was also very large (over 36,000 lines), and
// compiling it was very slow, and in fact was a significant
// bottleneck for incremental rebuilds. To address this, we now
// generate the file split across multiple shards, named
// VariableType_0.cpp and so on, which can be compiled in parallel.
//
// For ease of inspection and debugging, so that it's not necessary to
// go rooting around in multiple files, we also generate all the
// functions together in VariableTypeEverything.cpp. This generated
// file is only for convenience; it's not actually used in the
// build. If the file you're looking at now is one of the shards, you
// may want to switch over to the Everything variant to make you
// grepping smoother.

using namespace at;
using namespace torch::autograd::generated;
using namespace torch::autograd::generated::details;


namespace torch { namespace autograd {

namespace VariableType {
namespace{
  C10_UNUSED void reset_grad_accumulator(Variable & self) {
    AutogradMeta* meta = torch::autograd::impl::get_autograd_meta(self);
    if (meta != nullptr) {
      meta->grad_accumulator_.reset();
    }
  }
}

namespace {


${type_derived_method_definitions_AutogradNestedTensor}

${type_derived_method_definitions_AutogradCUDA}

${type_derived_method_definitions_Default}
}
}

namespace {

TORCH_LIBRARY_IMPL(aten, AutogradNestedTensor, m) {
${wrapper_registrations_AutogradNestedTensor}
}

TORCH_LIBRARY_IMPL(aten, AutogradCUDA, m) {
${wrapper_registrations_AutogradCUDA}
}

TORCH_LIBRARY_IMPL(aten, Autograd, m) {
${wrapper_registrations_Default}
}

}

}} // namespace torch::autograd
```
</details>

mikaylagawarecki replied:
Perhaps a variadic macro would work since there would be a variable number of arguments. However I'm not quite sure how to access individual arguments of __VA_ARGS__ (e.g. the Dispatch macros have `<macro>AND`, `<macro>AND2` etc.) which leads me to believe that unpacking a variable number of arguments in a single macro might be tricky (but I am not familiar with this at all so any thoughts would be appreciated!)

albanD replied:
I'm happy with the two step codegen here on my end.
Maybe we can leave a comment in the original template to mention this to avoid any confusion?

@bhosmer did you had specific concern about this practice?

bhosmer replied:
Just complexity - the codegen is already a very complex build-time process, and introducing the notion of chained codegen seems like a worrisome door to open, if there's a reasonable way to avoid it.

I haven't had bandwidth to think about what a solution might look like (but thanks a bunch Mikayla for the quick turnaround on a sample). But ofc if a solution is worse than the problem - or blocks Mikayla landing this - then I wouldn't insist. Gimme till like EOD to either suggest an alternative or give up 😛

bhosmer replied:
@mikaylagawarecki so the straightforward alternative is
* make `${type_derived_method_definitions}` iterate over all used keys and put the definitions out in one big slab
* make `${wrapper_registrations}` put out a `TORCH_LIBRARY_IMPL` block for each used key (and remove the literal `TORCH_LIBRARY_IMPL(...) {...}` text out of the template)

...but tbh it probably isn't worth the trouble - in particular the latter would need some nontrivial new plumbing.

Your current approach is a clever way to avoid all that, which is of course a plus from the POV of code reuse, churn etc. My worry was that having introduced the chained-codegen technique it might get used in less defensible situations, but that's on us to prevent down the line. Definitely not worth the extra work on this PR to avoid.

bhosmer reviewed (commented):
+1 to Alban's comment - really nice piece of work! Still need to review the last few files but meanwhile, here's some comments 😁  nothing major except (maybe) around the 2-hop codegen change

bhosmer reviewed (approved):
assuming @albanD doesn't spot anything I've missed, I say ship it

mikaylagawarecki committed:
Update on "Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"

`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


* __->__ #82801

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Enable codegen of per-dispatch key derivative formulas in derivatives.yaml"

`derivatives.yaml` can now take a `dispatch` entry which registers per-autograd dispatch key derivatives such as
```
name: foo(Tensor self, Tensor y) -> Tensor
dispatch:
  default:
    x: grad
    y: grad.expand(y.sizes())
  AutogradNestedTensor:
    x: grad
    y:  NestedTensor_foo_backward(grad, y)
output_differentiabilty: [True]
```

However the old schema where there is no `dispatch` entry is still supported.

Would greatly appreciate feedback on *how to improve the testing strategy* of this PR, currently have registered an aten test op in TestOps.cpp with dummy gradients in derivatives.yaml and have some tests in test_autograd.py:TestAutogradMultipleDispatch but I am not sure whether these are sufficiently rigorous.

Additionally, this PR also makes the assumption that sets like [VIEW_FUNCTIONS](https://github.com/pytorch/pytorch/blob/ff5399e528cf0be4c029a9396600a10b99eed668/tools/autograd/gen_inplace_or_view_type.py#L60) are per-native-function and not per-native-function-and-dispatch-key. I'm not sure whether this is necessarily the case, *would there ever be a situation where (e.g. a nested_tensor op is a view op but the aten function is not or vice versa?)*


* __->__ #82801

[ghstack-poisoned]

albanD reviewed (approved):
Looks great!

mikaylagawarecki commented:
@pytorchbot merge
